content/blog/pitfalls-of-safe-rust/index.md

```diff
@@ -0,0 +1,714 @@
+}
+```
+
+## Don't Accept Unbounded Input
+
+Protect Against Denial-of-Service Attacks with Resource Limits.
+These happen when you accept unbounded input, e.g. a huge request body
+which might not fit into memory.
+
+```rust
+// DON'T: Accept unbounded input
+fn process_request(data: &[u8]) -> Result<(), Error> {
+    let decoded = decode_data(data)?;  // Could be enormous!
+    // Process decoded data
+    Ok(())
+}
+```
+
+Instead, set explicit limits for your accepted payloads:
+
+```rust
+const MAX_REQUEST_SIZE: usize = 1024 * 1024;  // 1MB
+
+fn process_request(data: &[u8]) -> Result<(), Error> {
+    if data.len() > MAX_REQUEST_SIZE {
+        return Err(Error::RequestTooLarge);
+    }
+    
+    let decoded = decode_data(data)?;
+    
+    // For added safety, you can also check the decoded size 
+    if decoded.len() > MAX_REQUEST_SIZE * 2 {  
+        return Err(Error::DecodedDataTooLarge);
+    }
+    
+    Ok(())
+}
+```
+
+## Surprising Behavior of `Path::join` With Absolute Paths 
+
+If you use `Path::join` to join a relative path with an absolute path, it will silently replace the relative path with the absolute path.
+
+```rust
+use std::path::Path;
+
+fn main() {
+    let path = Path::new("/usr").join("/local/bin");
+    println!("{path:?}"); // Prints "/local/bin" 
+}
+```
+
+This is because `Path::join` will return the second path if it is absolute.
+
+I was not the only one who was confused by this behavior.
+Here's a [thread on the topic](https://users.rust-lang.org/t/rationale-behind-replacing-paths-while-joining/104288), which also includes an answer by [Johannes Dahlström](https://users.rust-lang.org/u/jdahlstrom/summary):
+
+> The behavior is useful because a caller [...] can choose whether it wants to
+> use a relative or absolute path, and the callee can then simply absolutize it by
+> adding its own prefix and the absolute path is unaffected which is probably what
+> the caller wanted. The callee doesn't have to separately check whether the path
+> is absolute or not.
+
+And yet, I still think it's a footgun.
+It's easy to overlook this behavior when you use user-provided paths.
+Perhaps `join` should return a `Result` instead?
+In any case, be aware of this behavior.
+
+
+## Check For Unsafe Code In Your Dependencies With `cargo-geiger`
+
+So far, we've only covered issues with your own code.
+For production code, you also need to check your dependencies.
+Especially unsafe code would be a concern.
+This can be quite challenging, especially if you have a lot of dependencies.
+
+[cargo-geiger](https://github.com/geiger-rs/cargo-geiger) is a neat tool that checks your dependencies for unsafe code.
+It can help you identify potential security risks in your project.
+
+```bash
+cargo install cargo-geiger
+cargo geiger
+```
+
+This will give you a report of how many unsafe functions are in your dependencies.
+If a non-sys level crate has a lot of unsafe code, that could be a red flag.
```

Comment:
Are crypto crates "non-sys level"? Not quite sure this is good advice. Or even valid in this blanket statement. Maybe ask readers to investigate the use of `unsafe` in that crate and judge if it's necessary or harmful for their application?

Reply:
True. Removed that sentence.